pixels/endpoints/general.py

```diff
@@ -0,0 +1,178 @@
+    r = requests.get("https://pixels.pythondiscord.com/pixels", headers=headers)
+    data = r.content
+
+    # have fun processing the returned data...
+    ```
+    """
+    # The cast to bytes here is needed by FastAPI ¯\_(ツ)_/¯
+    return Response(bytes(await request.state.canvas.get_pixels()),
+                    media_type="application/octet-stream")
+
+
+@router.get("/pixel", response_model=Pixel)
+@ratelimits.UserRedis(
+    requests=Ratelimits.GET_PIXEL_AMOUNT,
+    time_unit=Ratelimits.GET_PIXEL_RATE_LIMIT,
+    cooldown=Ratelimits.GET_PIXEL_RATE_COOLDOWN
+)
+async def get_pixel(x: int, y: int, request: Request) -> Pixel:
+    """
+    Get a single pixel given the x and y coordinates.
+
+    This endpoint requires an authentication token.
+    See [this page](https://pixels.pythondiscord.com/info/authentication)
+    for how to authenticate with the API.
+
+    #### Example Python Script
+    ```py
+    from dotenv import load_dotenv
+    from os import getenv
+    import requests
+
+    load_dotenv(".env")
+
+    token = getenv("TOKEN")
+
+    headers = {"Authorization": f"Bearer {token}"}
+
+    r = requests.get(
+        "https://pixels.pythondiscord.com/pixel",
+        headers=headers,
+        # Note: We're using query parameters to pass the coordinates, not the request body:
+        params={
+            "x": 87,
+            "y": 69
+        }
+    )
+
+    print("Here's the colour of the pixel:", r.json()["rgb"])
+    ```
+    """
+    if x >= Sizes.WIDTH or y >= Sizes.HEIGHT:
+        raise HTTPException(400, "Pixel is out of the canvas bounds.")
+    pixel_data = await request.state.canvas.get_pixel(x, y)
+
+    return Pixel(x=x, y=y, rgb=''.join(f"{x:02x}" for x in pixel_data))
+
+
+@router.put("/pixel", response_model=Message)
+@ratelimits.UserRedis(
+    requests=Ratelimits.PUT_PIXEL_AMOUNT,
+    time_unit=Ratelimits.PUT_PIXEL_RATE_LIMIT,
+    cooldown=Ratelimits.PUT_PIXEL_RATE_COOLDOWN
+)
+async def put_pixel(request: Request, pixel: Pixel) -> Message:
+    """
+    Override the pixel at the specified coordinate with the specified color.
+
+    This endpoint requires an authentication token.
+    See [this page](https://pixels.pythondiscord.com/info/authentication)
+    for how to authenticate with the API.
+
+    #### Example Python Script
+    ```py
+    from dotenv import load_dotenv
+    from os import getenv
+    import requests
+
+    load_dotenv(".env")
+
+    token = getenv("TOKEN")
+
+    headers = {"Authorization": f"Bearer {token}"}
+    data = {
+      "x": 80,
+      "y": 45,
+      "rgb": "00FF00"
+    }
+    # Remember, this is a PUT method.
+    r = requests.post(
+        "https://pixels.pythondiscord.com/pixel",
+        # Request body this time:
+        json=data,
+        headers=headers,
+    )
+    payload = r.json()
+
+    print(f"We got a message back! {payload['message']}")
+    ```
+    """
+    log.info(f"{request.state.user_id} is setting {pixel.x}, {pixel.y} to {pixel.rgb}")
```

Comment:
In all logger calls you can use the old-style formatting and pass args.
> Logs a message with level INFO on this logger. The msg is the message format string, and the args are the arguments which are merged into msg using the string formatting operator. (Note that this means that you can use keywords in the format string, together with a single dictionary argument.)

The benefit of this, is that if log level is above info, then no actual string formatting will be done. Compared to how it is right now, where Python will format this f-string and then pass it to info().

This may very much be an acceptable thing, as the following code is definitely not as readable.
```suggestion
    log.info("%s is setting %s, %s to %s", request.state.user_id, pixel.x, pixel.y, pixel.rgb)
```

The performance increase will be minor, but I thought I'd bring it up if you somewhat care about performance and commonly have log level above INFO.

Reply:
We run production with info, level logging, so I think we won't need this change.

But this reminds me that we should integrate sentry with this project, so warning+ logs get pushed there.